src/dune_pkg/solver_env.mli

```diff
@@ -1,84 +1,99 @@
+      (** Merge two environments returning an error if they both contain a
+          binding of the same variable to different values. *)
+      val union : t -> t -> (t, union_error) result
+    end
+  end
+end
 
-  module Bindings : sig
-    type sys_var := t
+module With_opt_build_config : sig
+  (** The default value for the "make" build config variable can't be computed
+      until after the workspace file has been parsed as it depends on the build
+      context. This types represents an environment where some build config
+      variables might be unset and can be converted into a full environment by
+      setting unset variables to their default values. *)
+  type t
 
-    (** A mapping from system environment variables to their values *)
-    type t
+  val default : t
 
-    val empty : t
+  val decode : t Dune_lang.Decoder.t
 
-    val set : t -> sys_var -> string -> t
+  val to_dyn : t -> Dyn.t
 
-    val get : t -> sys_var -> string option
+  val equal : t -> t -> bool
+end
 
-    type union_error =
-      [ `Var_in_both_with_different_values of sys_var * string * string ]
+(** A typed variable environment used by the dependency solver to evaluate
+    package filters. Opam packages can declare conditional dependencies on other
+    packages using a language made up of boolean operators and comparisons of
+    strings. Variables in this language can represent booleans and strings. *)
+type t
 
-    (** Merge two environments returning an error if they both contain a binding
-        of the same variable to different values. *)
-    val union : t -> t -> (t, union_error) result
-  end
-end
+(** Create a [t] from a [With_opt_build_config.t] filling in any unset fields
+    from the build config with default values. The [which_make] function is used
+    to look up the path of the `make` executable if the "make" build config
+    variable is unset. The [which_make] argument is ideally a function which
+    calls [Dune_rules.Expander.which_make] for the appropriate context. That
+    function can't be called directly from this module as it would create a
+    circular dependency. *)
+val with_default_values_for_unset_build_configs :
+     With_opt_build_config.t
+  -> which_make:(unit -> Path.t option Memo.t)
+  -> t Memo.t
 
-type t =
-  { flags : Flag.Set.t
-  ; sys : Sys_var.Bindings.t
-  }
+val sys_env : t -> Variable.Sys_env.Bindings.t
 
-val decode : t Dune_lang.Decoder.t
+val set_sys_env : t -> Variable.Sys_env.Bindings.t -> t
 
-val encode : t Dune_lang.Encoder.t
+(** Set all the flags to false *)
+val clear_flags : t -> t
 
-val to_dyn : t -> Dyn.t
+(** A human-readible summary of the variable environment *)
+val pp : t -> 'a Pp.t
 
-val equal : t -> t -> bool
+module Variable_value : sig
+  type t =
+    | Bool of bool
+    | String of string
+    | Unset_sys_env
+end
 
-val default : t
+(** Look up a variable's value by name *)
+val get_by_variable_name :
+  t -> string -> (Variable_value.t, [ `Unknown_variable ]) result
```

Comment:
Could you introduce a type for variable names?

Reply:
There is `Variable.t` inside solver_env.ml but it's not exposed. Think it would be clearer to use expose it and get rid of this helper function?

Reply:
We would still need to expose a way to convert strings (or I guess `OpamVariable.t`s) into the `Variable.t` type though. What's the purpose of avoiding raw strings? Is it just because the meaning of the string is not suggested by its type? Would replacing `string` in this api with `OpamVariable.t` be sufficient?

Reply:
There's two problem with using raw strings:

1. It makes reading the code harder than it needs be.
2. It becomes possible to skip the validation that is often necessary for these various types. For example, I'm pretty sure that an opam variable cannot contain the `'\0'` character. In some cases where we didn't add such validation, we ended up having to introduce painful workaround down the roads (see the code library names and module names).

Reply:
I would suggest against exposing `OpamVariable.t` in our API, as this ties us a bit stronger to OPAM and if we decide to change it later it would make things harder.